pkg/trace/obfuscate/sql_tokenizer.go

```diff
@@ -246,198 +247,210 @@ func (tkn *SQLTokenizer) skipBlank() {
 	for unicode.IsSpace(tkn.lastChar) {
 		tkn.next()
 	}
+	tkn.grab()
+}
+
+// myFastToUpper returns a copy of the byte slice s with all Unicode letters mapped to
+// their upper case. It uses the array pointed to by dst as the destination if possible.
+func myFastToUpper(src, dst []byte) []byte {
+	dst = dst[:0]
+	isASCII, hasLower := true, false
+	for i := 0; i < len(src); i++ {
+		c := src[i]
+		if c >= utf8.RuneSelf {
+			isASCII = false
+			break
+		}
+		hasLower = hasLower || ('a' <= c && c <= 'z')
+	}
+	if cap(dst) < len(src) {
+		dst = make([]byte, 0, len(src))
+	}
+	if isASCII { // optimize for ASCII-only byte slices.
+		if !hasLower {
+			// Just return a copy.
+			return append(dst, src...)
+		}
+		dst = dst[:len(src)]
+		//b := make([]byte, len(src))
+		for i := 0; i < len(src); i++ {
+			c := src[i]
+			if 'a' <= c && c <= 'z' {
+				c -= 'a' - 'A'
+			}
+			dst[i] = c
+		}
+		return dst
+	}
+	// This *could* be optimized, but it's an uncommon case.
+	return bytes.Map(unicode.ToUpper, src)
 }
 
 func (tkn *SQLTokenizer) scanIdentifier() (TokenKind, []byte) {
-	buffer := &bytes.Buffer{}
-	buffer.WriteRune(tkn.lastChar)
 	tkn.next()
-
 	for isLetter(tkn.lastChar) || isDigit(tkn.lastChar) || tkn.lastChar == '.' || tkn.lastChar == '*' {
-		buffer.WriteRune(tkn.lastChar)
 		tkn.next()
 	}
-	upper := bytes.ToUpper(buffer.Bytes())
+
+	t := tkn.grab()
+	// Space allows us to upper-case identifiers 200 bytes long or less without allocating.
+	var space [200]byte
```

Comment:
why 200?

Reply:
200 was chosen because it seemed reasonable. We may be able to find a more appropriate number. This is stack allocated as you mentioned, so we can get a larger array for the same (negligible) cost if we want. I'm not sure how common identifiers larger than 200 bytes are.

Reply:
I did a bit of research, and I think 256 bytes is a good number.
* Identifiers in mysql can be up to [256 characters](https://dev.mysql.com/doc/refman/8.0/en/identifier-length.html) (but only aliases on table names. Most are limited to 64 characters)
* For postgres, it's [63 characters](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS)
* For Oracle it's [128 characters](https://oracle-base.com/articles/12c/long-identifiers-12cr2#:~:text=In%20Oracle%20Database%2012cR2%2C%20the,identifiers%20is%20now%20128%20characters.)
* For MS SQL Server, the limit (as best I can tell) is [128 characters](https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms191240(v=sql.105)?redirectedfrom=MSDN#sysname)

As far as I can tell, ANSI/ISO SQL:2011 only specifies at least 30 characters.

256 bytes will cover almost all cases.
Characters and bytes aren't equivalent, but I think it's OK if queries with long, utf-filled aliases in mysql hit an extra alloc.